packages/server/message-store-api/service/MessageStoreService.go

```diff
@@ -203,15 +217,60 @@ func (s *MessageService) SaveMessage(ctx context.Context, input *msProto.InputMe
 
 	previewMessage := s.getMessagePreview(input)
 
-	senderType := s.commonStoreService.ConvertMSSenderTypeToEntitySenderType(input.SenderType)
-	if _, err := s.customerOSService.UpdateConversation(tenant, conversation.Id, *initiatorId, senderType, previewMessage); err != nil {
+	userIds := []string{}
+	contactIds := []string{}
+	for participantsIndex := range participants {
+		if participants[participantsIndex].Type == entity.CONTACT {
+			contactIds = append(contactIds, participants[participantsIndex].Id)
+		} else if participants[participantsIndex].Type == entity.USER {
+			userIds = append(userIds, participants[participantsIndex].Id)
+		}
+	}
+
+	lastSenderType := s.getSenderTypeStr(initiator)
+	if _, err := s.customerOSService.UpdateConversation(tenant, conversation.Id, initiator.Id, lastSenderType, contactIds, userIds, previewMessage); err != nil {
 		return nil, err
 	}
 
-	conversationEvent := s.saveConversationEvent(tenant, conversation, input, *initiatorId, senderType)
+	conversationEvent := s.saveConversationEvent(tenant, conversation, input, initiator)
 
 	return s.commonStoreService.EncodeMessageIdToMs(conversationEvent), nil
 }
+func (s *MessageService) getSenderTypeStr(initiator *Participant) string {
+	var lastSenderType = ""
+	if initiator.Type == entity.CONTACT {
+		lastSenderType = "CONTACT"
+	} else if initiator.Type == entity.USER {
+		lastSenderType = "USER"
+	}
+	return lastSenderType
+}
+
+func (s *MessageService) getThreadIdAndParticipantsFromMail(tenant string, threadId *string, participants []Participant, entityType entity.EventType, input *msProto.InputMessage) error {
```

Comment:
this won't work, passing an array creates a copy of the array,
and append creates a new array...

Reply:
Fixed